In [1]:
# Add your imports here
import pandas as pd
import numpy as np
import scipy as sp
from pyspark.sql import *
from pyspark.sql.types import *
import matplotlib.pyplot as plt
%matplotlib inline

from pyspark.sql import *
from pyspark.sql.functions import *
from pyspark import SparkContext

import re
import datetime
import os
import math

# create the session
spark = SparkSession.builder.getOrCreate()

# create the context
sc = spark.sparkContext

In [2]:
DATA_FOLDER = 'Data'
REVIEW_FOLDER_0 = DATA_FOLDER + '/amazon_reviews_us_Books_v1_00.tsv.gz'
REVIEW_FOLDER_1 = DATA_FOLDER + '/amazon_reviews_us_Books_v1_01.tsv.gz'
REVIEW_FOLDER_2 = DATA_FOLDER + '/amazon_reviews_us_Books_v1_02.tsv.gz'
COVER_FOLDER = DATA_FOLDER + '/book32-listing.csv'

In [4]:
#REVIEW_FOLDER_TEST = 'Data/amazon_reviews_us_Books_v1_00.tsv.gz'
#df_reviews_0 = spark.read.format('csv').option('delimiter', '\t').option('header', 'true').load(REVIEW_FOLDER_TEST)
#df_reviews_0.write.parquet('reviews_us_0.parquet', mode='overwrite')


In [7]:
# Add your code and description here
if os.path.exists(DATA_FOLDER+'/Reviews_us.parquet'):
    #load the dataset into a Spark dataframe
    df_reviews = spark.read.parquet(DATA_FOLDER+'/Reviews_us.parquet')
    #show the content (2 first rows)
    df_reviews.show(2)
else:
    #load the dataset into a Spark dataframe
    df_reviews_0 = spark.read.format('csv').option('delimiter', '\t').option('header', 'true').load(REVIEW_FOLDER_0)
    df_reviews_1 = spark.read.format('csv').option('delimiter', '\t').option('header', 'true').load(REVIEW_FOLDER_1)
    df_reviews_2 = spark.read.format('csv').option('delimiter', '\t').option('header', 'true').load(REVIEW_FOLDER_2)
    df_reviews = df_reviews_0.union(df_reviews_1).union(df_reviews_2)
    #show the content (2 first rows)
    df_reviews.show(2)
    #save it in the Parquet format
    df_reviews.write.parquet(DATA_FOLDER+'/Reviews_us.parquet', mode='overwrite')

+-----------+-----------+--------------+----------+--------------+--------------------+----------------+-----------+-------------+-----------+----+-----------------+--------------------+--------------------+-----------+
|marketplace|customer_id|     review_id|product_id|product_parent|       product_title|product_category|star_rating|helpful_votes|total_votes|vine|verified_purchase|     review_headline|         review_body|review_date|
+-----------+-----------+--------------+----------+--------------+--------------------+----------------+-----------+-------------+-----------+----+-----------------+--------------------+--------------------+-----------+
|         US|   25933450| RJOVP071AVAJO|0439873800|      84656342|There Was an Old ...|           Books|          5|            0|          0|   N|                Y|          Five Stars|I love it and so ...| 2015-08-31|
|         US|    1801372|R1ORGBETCDW3AI|1623953553|     729938122|      I Saw a Friend|           Books|          5|    

In [3]:
# Add your code and description here
if os.path.exists(DATA_FOLDER+'/Cover.parquet'):
    print("Loading from existing parquet...")
    #load the dataset into a Spark dataframe
    df_covers = spark.read.parquet(DATA_FOLDER+'/Cover.parquet')
    #show the content (2 first rows)
    df_covers.show(2)
else:
    print("Loading from csv...")
    #load the dataset into a Spark dataframe
    schema = StructType([
        StructField('product_id', StringType(), True), 
        StructField('filename', StringType(), True),
        StructField('image_url', StringType(), True),
        StructField('title', StringType(), True),
        StructField('author', StringType(), True),
        StructField('category_id', StringType(), True),
        StructField('category', StringType(), True)])
    df_covers = spark.read.format('csv').schema(schema).load(COVER_FOLDER)
    #show the content (2 first rows)
    df_covers.show(2)
    #save it in the Parquet format
    df_covers.write.parquet(DATA_FOLDER+'/Cover.parquet', mode='overwrite')

Loading from existing parquet...
+----------+--------------+--------------------+--------------------+------------------+-----------+----------------+
|product_id|      filename|           image_url|               title|            author|category_id|        category|
+----------+--------------+--------------------+--------------------+------------------+-----------+----------------+
|1285859480|1285859480.jpg|http://ecx.images...|      Microeconomics|     William Boyes|          2|Business & Money|
| 133885070|0133885070.jpg|http://ecx.images...|Economics Today: ...|Roger LeRoy Miller|          2|Business & Money|
+----------+--------------+--------------------+--------------------+------------------+-----------+----------------+
only showing top 2 rows



In [11]:
df_reviews_2 = spark.read.parquet(DATA_FOLDER+'/reviews_us_2.parquet')

In [12]:
df_reviews_2.count()

3105520

In [6]:
rdd_test = df_reviews_2.rdd

In [10]:
rdd_test.toDF()

DataFrame[marketplace: string, customer_id: string, review_id: string, product_id: string, product_parent: string, product_title: string, product_category: string, star_rating: string, helpful_votes: string, total_votes: string, vine: string, verified_purchase: string, review_headline: string, review_body: string, review_date: string]

In [13]:
df_test = df_reviews_2.withColumn('index', monotonically_increasing_id())


In [14]:
df_test_1 = df_test.sort('index').limit(100)

In [15]:
df_test_1.show(5)

Py4JJavaError: An error occurred while calling o88.showString.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 6 in stage 11.0 failed 1 times, most recent failure: Lost task 6.0 in stage 11.0 (TID 47, localhost, executor driver): java.lang.OutOfMemoryError: Java heap space
	at org.apache.parquet.hadoop.ParquetFileReader$ConsecutiveChunkList.readAll(ParquetFileReader.java:1064)
	at org.apache.parquet.hadoop.ParquetFileReader.readNextRowGroup(ParquetFileReader.java:698)
	at org.apache.spark.sql.execution.datasources.parquet.VectorizedParquetRecordReader.checkEndOfRowGroup(VectorizedParquetRecordReader.java:303)
	at org.apache.spark.sql.execution.datasources.parquet.VectorizedParquetRecordReader.nextBatch(VectorizedParquetRecordReader.java:258)
	at org.apache.spark.sql.execution.datasources.parquet.VectorizedParquetRecordReader.nextKeyValue(VectorizedParquetRecordReader.java:161)
	at org.apache.spark.sql.execution.datasources.RecordReaderIterator.hasNext(RecordReaderIterator.scala:39)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:109)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:186)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:109)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.scan_nextBatch_0$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$10$$anon$1.hasNext(WholeStageCodegenExec.scala:614)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.convert.Wrappers$IteratorWrapper.hasNext(Wrappers.scala:30)
	at org.spark_project.guava.collect.Ordering.leastOf(Ordering.java:628)
	at org.apache.spark.util.collection.Utils$.takeOrdered(Utils.scala:37)
	at org.apache.spark.sql.execution.TakeOrderedAndProjectExec$$anonfun$5.apply(limit.scala:150)
	at org.apache.spark.sql.execution.TakeOrderedAndProjectExec$$anonfun$5.apply(limit.scala:149)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$23.apply(RDD.scala:801)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$23.apply(RDD.scala:801)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:49)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:49)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:96)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
	at org.apache.spark.scheduler.Task.run(Task.scala:109)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:345)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1651)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1639)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1638)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1638)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:831)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:831)
	at scala.Option.foreach(Option.scala:257)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:831)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1872)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1821)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1810)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:642)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2034)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2055)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2074)
	at org.apache.spark.sql.execution.SparkPlan.executeTake(SparkPlan.scala:363)
	at org.apache.spark.sql.execution.CollectLimitExec.executeCollect(limit.scala:38)
	at org.apache.spark.sql.Dataset.org$apache$spark$sql$Dataset$$collectFromPlan(Dataset.scala:3278)
	at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2489)
	at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2489)
	at org.apache.spark.sql.Dataset$$anonfun$52.apply(Dataset.scala:3259)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:77)
	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3258)
	at org.apache.spark.sql.Dataset.head(Dataset.scala:2489)
	at org.apache.spark.sql.Dataset.take(Dataset.scala:2703)
	at org.apache.spark.sql.Dataset.showString(Dataset.scala:254)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.OutOfMemoryError: Java heap space
	at org.apache.parquet.hadoop.ParquetFileReader$ConsecutiveChunkList.readAll(ParquetFileReader.java:1064)
	at org.apache.parquet.hadoop.ParquetFileReader.readNextRowGroup(ParquetFileReader.java:698)
	at org.apache.spark.sql.execution.datasources.parquet.VectorizedParquetRecordReader.checkEndOfRowGroup(VectorizedParquetRecordReader.java:303)
	at org.apache.spark.sql.execution.datasources.parquet.VectorizedParquetRecordReader.nextBatch(VectorizedParquetRecordReader.java:258)
	at org.apache.spark.sql.execution.datasources.parquet.VectorizedParquetRecordReader.nextKeyValue(VectorizedParquetRecordReader.java:161)
	at org.apache.spark.sql.execution.datasources.RecordReaderIterator.hasNext(RecordReaderIterator.scala:39)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:109)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:186)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:109)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.scan_nextBatch_0$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$10$$anon$1.hasNext(WholeStageCodegenExec.scala:614)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.convert.Wrappers$IteratorWrapper.hasNext(Wrappers.scala:30)
	at org.spark_project.guava.collect.Ordering.leastOf(Ordering.java:628)
	at org.apache.spark.util.collection.Utils$.takeOrdered(Utils.scala:37)
	at org.apache.spark.sql.execution.TakeOrderedAndProjectExec$$anonfun$5.apply(limit.scala:150)
	at org.apache.spark.sql.execution.TakeOrderedAndProjectExec$$anonfun$5.apply(limit.scala:149)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$23.apply(RDD.scala:801)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$23.apply(RDD.scala:801)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:49)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:49)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:96)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
	at org.apache.spark.scheduler.Task.run(Task.scala:109)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:345)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)


----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 50294)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/socketserver.py", line 317, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/anaconda/lib/python3.6/socketserver.py", line 348, in process_request
    self.finish_request(request, client_address)
  File "/anaconda/lib/python3.6/socketserver.py", line 361, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/anaconda/lib/python3.6/socketserver.py", line 696, in __init__
    self.handle()
  File "/opt/spark/python/pyspark/accumulators.py", line 265, in handle
    poll(accum_updates)
  File "/opt/spark/python/pyspark/accumulators.py", line 238, in poll
    if func():
  File "/opt/spark/python/pyspark/accumulators.py", line 242, in accum_updates
    num_updates = read_int(self.rfile)
  File "/opt/spark/python/pyspark/serializers.py", line 69

In [12]:
df_test = df_reviews_2.withColumn('index', monotonically_increasing_id())
# sort ascending and take first 100 rows for df1
df_test_1 = df_test.sort('index').limit(100)

# sort descending and take 400 rows for df2
#df_test_2 = df_test.sort('index', ascending=False).limit(400)

In [13]:
df_test_1.show(5)

--- Logging error ---
ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49987)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-13-6d5c0f1ab7a1>", line 1, in <module>
    df_test_1.show(5)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above excepti

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49987)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-13-6d5c0f1ab7a1>", line 1, in <module>
    df_test_1.show(5)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception 

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49987)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-13-6d5c0f1ab7a1>", line 1, in <module>
    df_test_1.show(5)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception 

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49987)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-13-6d5c0f1ab7a1>", line 1, in <module>
    df_test_1.show(5)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception 

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49987)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-13-6d5c0f1ab7a1>", line 1, in <module>
    df_test_1.show(5)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception 

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49987)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-13-6d5c0f1ab7a1>", line 1, in <module>
    df_test_1.show(5)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception 

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49987)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-13-6d5c0f1ab7a1>", line 1, in <module>
    df_test_1.show(5)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception 

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49987)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-13-6d5c0f1ab7a1>", line 1, in <module>
    df_test_1.show(5)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception 

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49987)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-13-6d5c0f1ab7a1>", line 1, in <module>
    df_test_1.show(5)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception 

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49987)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-13-6d5c0f1ab7a1>", line 1, in <module>
    df_test_1.show(5)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception 

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49987)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-13-6d5c0f1ab7a1>", line 1, in <module>
    df_test_1.show(5)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception 

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49987)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-13-6d5c0f1ab7a1>", line 1, in <module>
    df_test_1.show(5)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception 

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49987)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-13-6d5c0f1ab7a1>", line 1, in <module>
    df_test_1.show(5)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception 

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49987)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-13-6d5c0f1ab7a1>", line 1, in <module>
    df_test_1.show(5)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception 

Py4JNetworkError: An error occurred while trying to connect to the Java server (127.0.0.1:49987)

In [8]:
df_joined = df_reviews_2.join(df_covers, ['product_id'])

In [9]:
df_joined.show(2)

--- Logging error ---
ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49914)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-9bd12c25d916>", line 1, in <module>
    df_joined.show(2)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exceptio

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49914)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-9bd12c25d916>", line 1, in <module>
    df_joined.show(2)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception o

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49914)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-9bd12c25d916>", line 1, in <module>
    df_joined.show(2)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception o

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49914)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-9bd12c25d916>", line 1, in <module>
    df_joined.show(2)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception o

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49914)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-9bd12c25d916>", line 1, in <module>
    df_joined.show(2)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception o

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49914)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-9bd12c25d916>", line 1, in <module>
    df_joined.show(2)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception o

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49914)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-9bd12c25d916>", line 1, in <module>
    df_joined.show(2)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception o

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49914)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-9bd12c25d916>", line 1, in <module>
    df_joined.show(2)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception o

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49914)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-9bd12c25d916>", line 1, in <module>
    df_joined.show(2)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception o

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49914)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-9bd12c25d916>", line 1, in <module>
    df_joined.show(2)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception o

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49914)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-9bd12c25d916>", line 1, in <module>
    df_joined.show(2)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception o

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49914)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-9bd12c25d916>", line 1, in <module>
    df_joined.show(2)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception o

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49914)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-9bd12c25d916>", line 1, in <module>
    df_joined.show(2)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception o

ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:49914)
Traceback (most recent call last):
  File "/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-9bd12c25d916>", line 1, in <module>
    df_joined.show(2)
  File "/opt/spark/python/pyspark/sql/dataframe.py", line 350, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/anaconda/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/opt/spark/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/anaconda/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
    format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: <unprintable Py4JJavaError object>

During handling of the above exception, another exception o

Py4JNetworkError: An error occurred while trying to connect to the Java server (127.0.0.1:49914)

In [10]:
print(df_reviews.count(), df_joined.count())

10319090 1423654


We read the dataset containing the reviews. We have to specify the columns names because some rows have a higher count of cell than others. We want to see what is contained in those rows. 

In [2]:
cols = ["marketplace",
        "customer_id",
        "review_id",
        "product_id",
        "product_parent",
        "product_title",
        "product_category",
        "star_rating",
        "helpful_votes",
        "total_votes",
        "vine",
        "verified_purchase",
        "review_headline",
        "review_body",
        "review_date", 
        "undefined_1",
        "undefined_2",
        "undefined_3",
        "undefined_4",
        "undefined_5",
        "undefined_6",
        "undefined_7"
       ]

In [3]:
#df_reviews = pd.read_table("amazon_reviews_us_Books_v1_00.tsv", error_bad_lines=False)
df_reviews = pd.read_csv(REVIEW_FOLDER, names=cols, sep='\t')

NameError: name 'pd' is not defined

The warning simply indicates that some columns do not have the same dtype for each row, this will be handled later by cleaning the data. 

So first we check the rows with a higher count of cells.

In [92]:
df_reviews[df_reviews["undefined_1"].notnull()]

marketplace customer_id       review_id  product_id product_parent  \
3523             US    45223684  R1NHKVD5LMP2JY  1503551814      353548609   
5281             US    18701079  R1946J1OJKC1K2  0743292472        7685359   
20477            US    40945620  R33MH6R1TZWIGR  1452572747      523136939   
25894            US    22921197  R1OPQAX1D7DWB8  0984755101      632337521   
27015            US    44087253  R24H2K78E883A0  1616496169      709191612   
59797            US    44250036   RYNEKW5P1Z3AG  0385262078      887509044   
69197            US    50641462  R2VI5KEAQR2P23  0545436826      247155770   
71952            US    43676679  R2F0SUDYHR79JQ  1844094642      766424526   
78719            US     4881222  R21NUSCF1WC1I8  1607512572      236705009   
81299            US     1148395  R1UH3R71GX97AR  1502716984      563128335   
87682            US    50861690  R2Z0YTRQKVFHSF  0807403083      147336042   
90515            US    28736875   RVFNQHOLE58A0  0806120940      661253970   
91146            US    15629461    R4U6722KCUT6  1592408419      543673826   
108944           US     8612032  R2V2ES66TGP7RN  1460268601      273651829   
118288           US    50953132  R2YWTZB0TCB920  1439132852      537451993   
124666           US    38871814  R1CQPRDZKZZKKN  097860783X      764135422   
125540           US    26965291  R33FTU1VADZM18  1591843820      618286357   
144114           US    19357930  R370TRRWYXD4HH  0743230418      757722318   
167695           US    49384329  R2ZVVLUZWJ1EWP  1502328585      460342874   
179855           US    51750629  R1919SOCLXFHH3  1456550144      598402352   
189711           US    41990563   RBWE5V226JS4A  0762454865      402819053   
196487           US    20804529  R1W0RQ0ETQT7V9  1475812140      387091568   
232131           US    26084626   RX8T8JKA1RCM4  0062333933      223598514   
241733           US    36457256  R32B42EMWA43S7  1400076560      204550997   
269922           US    17688934   R70Y8Z54OL2OW  0316210463      723771425   
279157           US    12975716  R3KDGN68ECWPFL  1285755618      475852195   
283356           US    15648568  R22WHMC556RKAR  386521584X      467228901   
285766           US    11380786  R12VP0F0Q16NMF  0971339309      192326528   
308165           US    41713493  R1EW4GKVIF0ALE  1503019500      768980686   
315198           US    45994291   RK3U2RRENGDFM  1511510056      439740394   
...             ...         ...             ...         ...            ...   
4449652          US     2484994  R3K2FPQZ7X8UVP  1497437016      519821647   
4459301          US    40203782  R1SILLK0ZNLGSD  1434238822      755154206   
4477212          US    14426796  R3HBOV96BGZSP2  0870717421      769808046   
4500862          US    13558114  R2F60Z8M7MYER2  038573252X      305985379   
4549885          US    46366696  R2PAOB044HM77A  1492927449      113630502   
4552287          US     4274094   RXS1XC8LMKG1Y  0823422437      304889763   
4558858          US    27197209  R2GNMOUPSY80XK  1611861314      616413203   
4562250          US    49569659   R7N4GP96VKLSA  9231039911      306686978   
4567913          US    52956445  R13129ZA040001  0470376295      280360616   
4633360          US    37024654   RGSFJRNLHB30C  0199654441      755550981   
4640816          US    19973002  R2OQPGHTYWVCZ2  0991067886      313246709   
4644916          US    30817773  R1WAHI6773B6KO  1497511135      946709255   
4671852          US    32651993  R3UV6QM6ZDKGYU  0385184581      212671778   
4701153          US     6071575   R9V19CKJTF8GG  055380653X      483985572   
4721069          US    38970700   RRGRPELJQGFI6  0822210738      994387679   
4731355          US    36857681  R3S1ZPUV914YNM  0761536434      630858517   
4744854          US    44416620  R1ZCFR2H9H98QN  1402720602       84227631   
4939719          US    47464628  R1RTM2W1I88ZGU  031074265X      972127661   
5184882          US    16100572  R10J8LHGN2V8TO  1935166301      218847016   
5279519          US    2384107

In [93]:
len(df_reviews[df_reviews["undefined_1"].notnull()])

326

What is contained in the undefined columns is a bit weird, it's like we have more reviews, but they are not really related to the standard review column. For example, in the second line of the output, the review is "Was a great read..." and then we have "NOT PLEASED WITH THE VENDORS AT ALL" etc.

Since these lines are not clear concerning the reviews, and since we only have 326 of such rows over 6 millions, we decide to drop these rows. 

In [ ]:
df_reviews = df_reviews[df_reviews[df_reviews["undefined_1"].isnull()]]

In [91]:
df_reviews[df_reviews["helpful_votes"].astype('str').apply(
    lambda x: int(x) if x.isnumeric() else np.nan) > 100]

marketplace customer_id       review_id  product_id product_parent  \
2093              US    15365576  R38I0D0J8O6LB6  048680318X      338332455   
7914              US    43484359  R2T373PSIAQURD  162315569X      635788774   
9107              US    30673354  R1F7PAD7F00VUI  1119057027      373857330   
11505             US    50489373  R1HO6TV1G31KA9  1623363586      465642569   
11998             US    13098534   RE2KQV5I1AS5Z  1780287518      268093982   
16284             US    52751117  R3E34ZT1227WIF  1477829407      529331494   
19789             US    45943820  R2JAD68WXL1PSY  0007525540      669379389   
21361             US    52936144  R34VXM3EJDS78Z  1594634718      892490739   
34930             US    13768935  R2JF5N8P1JTJVE  0486799026      424027076   
35222             US    29532845  R3SVHW4TTDH9KF  0812995821      604086967   
36808             US    22335386  R1FJJTZWM2PA6E  1617292397      243805263   
36948             US    53043025   RCLSDAIJR6FUG  0804136696      883209083   
38551             US    16769854  R3LP3EL4H7P9WD  0545795664      598859984   
43273             US    51415904  R38YGIBKP8X1WN  B00QAW07NW      841757074   
48782             US    50082032   R9T2KZNYUTCSR  0718181239      145399331   
57591             US    34172652  R1G3NSRI8ANN0R  1250067189      325106175   
58269             US    23846195  R2N47RIBWYXTU5  1433688670      109159201   
59577             US    48428870  R1IHJRB3IBLEPM  1454709138      283295941   
61090             US    51210331   RHX42CST5IFQF  1250022088      918932098   
64019             US    19355425  R3DYNWGGNGD2T9  099112720X      789387828   
65249             US    41411148  R3EBTH6UEIIZOT  193770226X      527559049   
65328             US    44604397  R3ENZXVUURS5HT  0385376715      812707771   
100855            US    48428870  R2GMB63H9MMH22  1454709286       98658686   
100878            US    51071607   R1G69NBCA01NO  0718031822      227928809   
101044            US    15365576  R3Q3SAZ0JXXM7G  0785832882      901228817   
110695            US    50739850   R4AF0D0ZQ1AJ0  0425017567      739394524   
111163            US    52687966  R2NI1E9ABPL0ZI  161243438X       54594476   
111173            US    51639969  R2MI4U9393MUWT  1119038596      517619840   
125210            US    16704405   RPCMO66E3M94O  0692239006      106288063   
127359            US    50392120  R214NJTRQ0NLE0  0778317706      204702637   
...              ...         ...             ...         ...            ...   
10227940          US    26166800  R16R475IXML1VT  1932735410      828399386   
10228155          US    44659482   RRIBO10N5ILPF  1594035725      512398500   
10228276          US    51889003  R1S4DF1P5CSQXR  0142419389      354462662   
10228693          US    49841439   RYHAD4N6R19LG  1937007448      512918432   
10228857          US    34815346  R2KOSWAM5J6CR2  1612130283      600633062   
10228886          US    26141371   R80AID6ZU9K25  0960446400      729379935   
10229129          US    44153355   RDEZDP0Q1R24D  141659907X      170659637   
10229187          US    46491010  R3JX1IN1I6CAEF  0770435629      243194079   
10229658          US    13189282  R380W2QI0B4O5Q  0465031331      164973150   
10229686          US    52830381  R2ATIJCX4DJWBB  1400069289      136914857   
10229773          US    24848903  R376TDKPMZ43FM  0761155589      818975183   
10230312          US    13161706    RIQBHKURZ75X  0465031331      164973150   
10230574          US    35575529  R2GK2V9178PRVH  0062059939      701141905   
10230915          US    51910704  R18CJB9URMZ2QK  0679405070        1031374   
10232488          US    40289210   RFBPIXT0QEBQT  159420327X      517011613   
10232491          US    40988773   RUEONMCOYS5RJ  0983490716      623999156   
10232649          US    50682072  R111CY5T3KVFCR  0618982507      989177484   
10232708          US    49332731  R2BIDBQXD75ILL  1436338999      571592578   
10233567          US    37701650  R2NXKX6LE8888I  159420327

In [73]:
df_reviews[df_reviews["undefined_1"].notnull()].iloc[:, 10:]

vine verified_purchase  \
3523       N                 Y   
5281       N                 Y   
20477      N                 Y   
25894      N                 Y   
27015      N                 Y   
59797      N                 Y   
69197      N                 Y   
71952      N                 Y   
78719      N                 Y   
81299      N                 Y   
87682      N                 Y   
90515      N                 Y   
91146      N                 Y   
108944     N                 N   
118288     N                 Y   
124666     N                 N   
125540     N                 N   
144114     N                 Y   
167695     N                 N   
179855     N                 Y   
189711     N                 N   
196487     N                 N   
232131     N                 Y   
241733     N                 N   
269922     N                 Y   
279157     N                 N   
283356     N                 Y   
285766     N                 N   
308165     N                 N   
315198     N                 N   
...      ...               ...   
4449652    N                 N   
4459301    N                 N   
4477212    N                 Y   
4500862    N                 Y   
4549885    N                 Y   
4552287    N                 N   
4558858    N                 N   
4562250    N                 N   
4567913    N                 Y   
4633360    N                 Y   
4640816    N                 Y   
4644916    N                 N   
4671852    N                 N   
4701153    N                 N   
4721069    N                 Y   
4731355    N                 Y   
4744854    N                 Y   
4939719    N                 N   
5184882    N                 Y   
5279519    N                 N   
5303009    N                 Y   
5659631    N                 Y   
5893491    N                 Y   
5906172    N                 Y   
6053733    N                 N   
6250799    N                 N   
6525509    N                 N   
6704024    N                 Y   
7724707    N                 N   
8647348    N                 N   

                                           review_headline review_body  \
3523      She is a fine storyteller and has a message c...       Books   
5281      was a great read. Quick-moving with sharp int...       Books   
20477    This wonderful book is thoughtful and well-wri...       Books   
25894     this is the best book on health and wellness ...       Books   
27015    Oh would some Power the gift give us, To ...\t...       Books   
59797     One of the most important of all American men...       Books   
69197     It is a bunch of sweet musings about friendsh...       Books   
71952     I intensely dislike the psychology perspectiv...       Books   
78719    discriptive of characters w/repetitive plot di...       Books   
81299     is a complete waste of time and money\tthis &...       Books   
87682     his choice of examples is excellent, his comm...       Books   
90515    The raiders carried off such great numbers of ...       Books   
91146    If you want to be happy in life\tAll hyped up ...       Books   
108944   The New Arcadia offers a vital and compelling ...       Books   
118288   Guns good! Guvmint bad\tsorry two stars. I cou...       Books   
124666    If you're looking for an introduction to the ...       Books   
125540    This was a great book. Remember two things\tI...       Books   
144114   I WARNED YOU!\tSTOOPIDIST, MOST IDIOTIC, NON-S...       Books   
167695    I took Karen's class at Art Unraveled and I w...       Books   
179855   damn fine work, sir\tSuperb! Sometimes you rea...       Books   
189711    I can see why he recommended this book\tAntho...       Books   
196487   Awesome book!... a refreshing take on ...\t&#3...       Books   
232131    They all go a long way toward helping us unde...       Books   
241733   The terror that permeated our world like a mys...       Books   
269922   I really enjoyed this book\tFrom my daug

In [53]:
test3[test3["product_id"] == "014006768X"].head()

marketplace  customer_id       review_id  product_id  product_parent  \
33670            US     31269475  R2VT0IUKEZ1IZI  014006768X        98817612   
566408           US     10029757   RK2X98AB8WAVH  014006768X        98817612   
776844           US     16010139  R3SI9L25CWYZ4I  014006768X        98817612   
885302           US     16587837  R3TALFOQHVFEFN  014006768X        98817612   
1791515          US     33892769  R19U49VOZRB3ZN  014006768X        98817612   

                     product_title product_category star_rating  \
33670    The First Rumpole Omnibus            Books           5   
566408   The First Rumpole Omnibus            Books           5   
776844   The First Rumpole Omnibus            Books           5   
885302   The First Rumpole Omnibus            Books           5   
1791515  The First Rumpole Omnibus            Books           5   

         helpful_votes  total_votes vine verified_purchase  \
33670              0.0          1.0    N                 Y   
566408             0.0          1.0    N                 Y   
776844             1.0          1.0    N                 Y   
885302             1.0          1.0    N                 N   
1791515            0.0          1.0    N                 Y   

                                           review_headline  \
33670                                           Great fun!   
566408                                          Entertaing   
776844   Fortunately many of these stories were made in...   
885302                                  Humanity on parade   
1791515                                         Five Stars   

                                               review_body review_date  
33670    Mortimer and his brain-child, Rumpole are both...  2015-08-28  
566408   I love it the only two author's that could bri...  2015-07-10  
776844   Fortunately many of these stories were made in...  2015-06-19  
885302   I was sitting alone at the bar of a coastal sa...  2015-06-09  
1791515                            Can't beat these books1  2015-03-16

In [25]:
test3.head()

marketplace  customer_id       review_id  product_id  product_parent  \
0          US     25933450   RJOVP071AVAJO  0439873800        84656342   
1          US      1801372  R1ORGBETCDW3AI  1623953553       729938122   
2          US      5782091   R7TNRFQAOUTX5  142151981X       678139048   
3          US     32715830  R2GANXKDIFZ6OI  014241543X       712432151   
4          US     14005703  R2NYB6C3R8LVN6  1604600527       800572372   

                                  product_title product_category star_rating  \
0  There Was an Old Lady Who Swallowed a Shell!            Books           5   
1                                I Saw a Friend            Books           5   
2                          Black Lagoon, Vol. 6            Books           5   
3                                     If I Stay            Books           5   
4                       Stars 'N Strips Forever            Books           5   

   helpful_votes  total_votes vine verified_purchase  \
0            0.0          0.0    N                 Y   
1            0.0          0.0    N                 Y   
2            0.0          0.0    N                 Y   
3            0.0          0.0    N                 N   
4            2.0          2.0    N                 Y   

                                     review_headline  \
0                                         Five Stars   
1  Please buy "I Saw a Friend"! Your children wil...   
2                                      Shipped fast.   
3                                         Five Stars   
4                                         Five Stars   

                                         review_body review_date  
0                 I love it and so does my students!  2015-08-31  
1  My wife and I ordered 2 books and gave them as...  2015-08-31  
2  Great book just like all the others in the ser...  2015-08-31  
3                                       So beautiful  2015-08-31  
4  Enjoyed the author's story and his quilts are ...  2015-08-31

In [58]:
test = test3["star_rating"].astype('str').apply(lambda x: x if x.isnumeric() else np.nan).dropna()


0           5
1           5
2           5
3           5
4           5
5           2
6           5
7           5
8           5
9           5
10          5
11          5
12          5
13          4
14          3
15          5
16          5
17          5
18          5
19          5
20          5
21          5
22          5
23          5
24          5
25          5
26          5
27          5
28          3
29          5
           ..
10236820    5
10236821    5
10236822    5
10236823    5
10236824    5
10236825    5
10236826    3
10236827    5
10236828    5
10236829    4
10236830    5
10236831    4
10236832    5
10236833    5
10236834    5
10236835    5
10236836    5
10236837    4
10236838    5
10236839    3
10236840    5
10236841    1
10236842    3
10236843    4
10236844    5
10236845    5
10236846    4
10236847    4
10236848    5
10236849    5
Name: star_rating, Length: 8205186, dtype: object

In [ ]:
test3[test3["star_rating"] == "2015-08-04"]

In [54]:
test3[test3["star_rating"] == "2015-08-04"]

marketplace  customer_id       review_id  product_id  product_parent  \
299292          US     42182727  R3IUDW7J4XLFHV  B003V8JQAI       198998645   
303458          US     52159729  R2S7P4O3PDJKSL  0988710501       101892350   

                                            product_title  \
299292  Dare to Be Great Ms. Caucus a Doonebury Book\t...   
303458  Prostate Cancer Breakthroughs 2014: New Tests,...   

                                         product_category star_rating  \
299292  This is a very objective, highly interesting, ...  2015-08-04   
303458  Harry Turtledove is an award-winning writer of...  2015-08-04   

        helpful_votes  total_votes vine verified_purchase review_headline  \
299292            NaN          NaN  NaN               NaN             NaN   
303458            NaN          NaN  NaN               NaN             NaN   

       review_body review_date  
299292         NaN         NaN  
303458         NaN         NaN

In [14]:
test3.loc[:, "star_rating"].value_counts()

5             5829705
5             1365286
4             1315264
3              567500
1              344802
4              282042
2              278990
3              121547
1               73240
2               58381
2015-08-04          2
2015-03-27          2
2015-06-11          1
2015-02-25          1
2014-07-24          1
2015-05-02          1
2015-07-06          1
2014-03-04          1
2015-05-27          1
2013-04-17          1
2015-04-05          1
2014-11-24          1
2013-05-28          1
2014-10-13          1
2015-06-05          1
2015-07-12          1
2015-05-20          1
2015-08-25          1
2014-12-01          1
2014-11-12          1
2014-09-29          1
2015-06-14          1
2015-04-17          1
2014-08-25          1
2013-08-24          1
2015-08-08          1
2015-05-29          1
2015-08-31          1
2013-08-03          1
2015-07-15          1
2015-05-24          1
Name: star_rating, dtype: int64

In [17]:
ratings = test3.loc[:, "star_rating"]

In [23]:
(ratings == 5).sum()

5829705

In [24]:
(ratings == "5").sum()

1365286

In [31]:
covers = pd.read_csv("book32-listing.csv", encoding = "ISO-8859-1", header=None)

In [32]:
columns = ["[AMAZON INDEX (ASIN)}","[FILENAME]","[IMAGE URL]","[TITLE]","[AUTHOR]","[CATEGORY ID]","[CATEGORY]"]

In [34]:
covers.columns = columns

In [37]:
covers.loc[10010:10020, :]

[AMAZON INDEX (ASIN)}      [FILENAME]  \
10010             385353596  0385353596.jpg   
10011            014006768X  014006768X.jpg   
10012            098544648X  098544648X.jpg   
10013             440243823  0440243823.jpg   
10014             312033567  0312033567.jpg   
10015            038550120X  038550120X.jpg   
10016             385339666  0385339666.jpg   
10017            1594136637  1594136637.jpg   
10018            B00CL62PS6  B00CL62PS6.jpg   
10019             316203440  0316203440.jpg   
10020             446612871  0446612871.jpg   

                                             [IMAGE URL]  \
10010  http://ecx.images-amazon.com/images/I/51iNYVpL...   
10011  http://ecx.images-amazon.com/images/I/51pTgRqM...   
10012  http://ecx.images-amazon.com/images/I/51CYTK4O...   
10013  http://ecx.images-amazon.com/images/I/51vgdKFN...   
10014  http://ecx.images-amazon.com/images/I/41G8381G...   
10015  http://ecx.images-amazon.com/images/I/51NAC46Q...   
10016  http://ecx.images-amazon.com/images/I/519R15FS...   
10017  http://ecx.images-amazon.com/images/I/51LWl64G...   
10018  http://ecx.images-amazon.com/images/I/51vdTEk4...   
10019  http://ecx.images-amazon.com/images/I/516H8c-H...   
10020  http://ecx.images-amazon.com/images/I/615gMuMj...   

                                                 [TITLE]          [AUTHOR]  \
10010                        The Jezebel Remedy: A novel      Martin Clark   
10011                          The First Rumpole Omnibus     John Mortimer   
10012                                  Cage Without Bars    Gerald Neufeld   
10013                             The Associate: A Novel      John Grisham   
10014                                Anatomy of a Murder     Robert Traver   
10015                                    A Painted House      John Grisham   
10016                                        The Chamber      John Grisham   
10017  Tuesdays Child (Thorndike Press Large Print Core)     Fern Michaels   
10018            An Innocent Client: Joe Dillard, Book 1       Scott Pratt   
10019  The Lincoln Lawyer Novels: The Lincoln Lawyer,...  Michael Connelly   
10020                                       Sudden Death   David Rosenfelt   

       [CATEGORY ID]                    [CATEGORY]  
10010             17  Mystery, Thriller & Suspense  
10011             17  Mystery, Thriller & Suspense  
10012             17  Mystery, Thriller & Suspense  
10013             17  Mystery, Thriller & Suspense  
10014             17  Mystery, Thriller & Suspense  
10015             17  Mystery, Thriller & Suspense  
10016             17  Mystery, Thriller & Suspense  
10017             17  Mystery, Thriller & Suspense  
10018             17  Mystery, Thriller & Suspense  
10019             17  Mystery, Thriller & Suspense  
10020             17  Mystery, Thriller & Suspense